src/TypeSpec.Extension/Emitter.Csharp/src/lib/confidentLevels.ts

```diff
@@ -0,0 +1,78 @@
+import {
+    InputDictionaryType,
+    InputListType,
+    InputModelType,
+    InputType,
+    isInputDictionaryType,
+    isInputListType,
+    isInputModelType
+} from "../type/inputType.js";
+
+export function getConfident(
+    type: InputType,
+    visitedModels: Set<InputModelType>
+): boolean {
+    if (isInputModelType(type)) {
+        return getModelConfident(type, visitedModels);
+    } else if (isInputListType(type)) {
+        return getListConfident(type, visitedModels);
+    } else if (isInputDictionaryType(type)) {
+        return getDictionaryConfident(type, visitedModels);
+    }
+
+    return type.IsConfident ?? true; // by default they are confident
+}
+
+function getModelConfident(
+    type: InputModelType,
+    visitedModels: Set<InputModelType>
+): boolean {
+    if (visitedModels.has(type)) return true; // this means this has been calculated before or being calculated right now, return
+    visitedModels.add(type);
+
+    if (type.IsConfident !== null) return type.IsConfident;
+
+    const isBaseLowConfident = type.BaseModel
+        ? getModelConfident(type.BaseModel, visitedModels)
+        : true;
+
+    let isModelLowConfident = isBaseLowConfident;
```

Comment:
I think you don't save the reaults of properties right? If so we could short circuit here.

Reply:
This function is not only calculating this value for this model, it is also "propagating" the calculation to every model it is referencing, therefore we should not short circuit here to ensure every type of the property has been gone through.
The property itself has a cache to ensure there is no duplicated calculations.

Reply:
Make sense, we should short circuit here